chemprop/v2/data/utils.py

```diff
@@ -2,65 +2,253 @@
+    :param num_folds: Number of folds to create (only needed for "cv" split type).
+    :param args: A :class:`~chemprop.args.TrainArgs` object.
+    :param logger: A logger for recording output.
+    :return: A tuple of :class:`~chemprop.data.MoleculeDataset`\ s containing the train,
+             validation, and test splits of the data.
+    """
     if not (len(sizes) == 3 and np.isclose(sum(sizes), 1)):
         raise ValueError(f"Invalid train/val/test splits! got: {sizes}")
 
-    n_train, n_val, n_test = [int(size * len(data)) for size in sizes]
-
-    if split == "random":
-        idxs = np.arange(len(data))
-        np.random.shuffle(idxs)
-
-        train = [data[i] for i in idxs[:n_train]]
-        val = [data[i] for i in idxs[n_train : n_train + n_val]]
-        test = [data[i] for i in idxs[n_train + n_val :]]
-    elif split == "scaffold":
-        scaffold2idxs = defaultdict(set)
-        for i, d in enumerate(data):
-            scaffold = MurckoScaffold.MurckoScaffoldSmiles(mol=d.mol)
-            scaffold2idxs[scaffold].add(i)
-
-        big_index_sets = []
-        small_index_sets = []
-        for idxs in scaffold2idxs.values():
-            if len(idxs) > n_val / 2 or len(idxs) > n_test / 2:
-                big_index_sets.append(idxs)
+    random = Random(seed)
+
+    if args is not None:
+        folds_file, val_fold_index, test_fold_index = (
+            args.folds_file,
+            args.val_fold_index,
+            args.test_fold_index,
+        )
+    else:
+        folds_file = val_fold_index = test_fold_index = None
+
+    # typically include a validation set
+    include_val = True
+    split_fun = train_val_test_split
+    mol_split_fun = train_val_test_split_molecules
+    # default sampling arguments for astartes sampler
+    astartes_kwargs = dict(
+        train_size=sizes[0], test_size=sizes[2], return_indices=True, random_state=seed
+    )
+    # if no validation set, reassign the splitting functions
+    if not sizes[1]:
+        include_val = False
+        split_fun = train_test_split
+        mol_split_fun = train_test_split_molecules
+    else:
+        astartes_kwargs["val_size"] = sizes[1]
+
+    train, val, test = None, None, None
+    if split == "crossval":
+        index_set = args.crossval_index_sets[args.seed]
+        data_split = []
+        for split in range(3):
+            split_indices = []
+            for index in index_set[split]:
+                with open(os.path.join(args.crossval_index_dir, f"{index}.pkl"), "rb") as rf:
+                    split_indices.extend(pickle.load(rf))
+            data_split.append([data[i] for i in split_indices])
+        train, val, test = tuple(data_split)
+
+    elif split in {"cv", "cv-no-test"}:
+        if num_folds <= 1 or num_folds > len(data):
+            raise ValueError(
+                "Number of folds for cross-validation must be between 2 and len(data), inclusive."
+            )
+
+        random = Random(0)
```

Comment:
Why is the `seed` set to 0 but not using the param `seed`?

Reply:
This has been removed in later commits.